File: 1047. Remove All Adjacent Duplicates in String/remove_duplicates.py
'''
LeetCode problem 1047. Remove All Adjacent Duplicates In String
https://leetcode.com/problems/remove-all-adjacent-duplicates-in-string/

You are given a string `s` consisting of lowercase English letters.
A duplicate removal consists of choosing two adjacent and equal letters and removing them.
We repeatedly make duplicate removals on `s` until we no longer can.
Return the final string after all such duplicate removals have been made.
It can be proven that the answer is unique.
'''


def remove_duplicates(s: str) -> str:
    '''
    Time complexity: O(n)
    Space complexity: O(n)
    '''
    if len(s) < 2:
        return s
    chars = [s[0]]
    i = 1
    while True:
        if len(chars) >= 1 and chars[-1] == s[i]:
            chars.pop()
        else:
            chars.append(s[i])
        i += 1
        if i == len(s):
            break
    return ''.join(chars)

File: 26. Remove Duplicates from Sorted Array/cs/RemoveDuplicates/RemoveDuplicates.cs
// LeetCode problem: 26. Remove Duplicates from Sorted Array
// https://leetcode.com/problems/remove-duplicates-from-sorted-array/description/

namespace LeetCode.RemoveDuplicates;

public class Solution
{
    public int RemoveDuplicates(int[] nums)
    {
        if (nums.Length == 0)
        {
            return 0;
        }

        int insertIndex = 1;
        for (int i = 1; i < nums.Length; i++)
        {
            if (nums[i - 1] != nums[i])
            {
                nums[insertIndex] = nums[i];
                insertIndex += 1;
            }
        }

        return insertIndex;
    }
}